tfhe/src/integer/server_key/radix/even_odd.rs

```diff
@@ -0,0 +1,92 @@
+use crate::integer::{BooleanBlock, IntegerRadixCiphertext, ServerKey};
+
+impl ServerKey {
+    /// Returns an encryption of true if the value is even
+    ///
+    /// ct is not required to have clean carries
+    pub fn unchecked_is_even<T>(&self, ct: &T) -> BooleanBlock
+    where
+        T: IntegerRadixCiphertext,
+    {
+        ct.blocks().first().map_or_else(
+            || {
+                // Interpret empty as being 0, which is even
+                self.create_trivial_boolean_block(true)
+            },
+            |first_block| {
+                let lut = self
+                    .key
+                    .generate_lookup_table(|block| u64::from((block & 1) == 0));
+                let result = self.key.apply_lookup_table(first_block, &lut);
+                BooleanBlock::new_unchecked(result)
+            },
+        )
+    }
+
+    /// Returns an encryption of true if the value is odd
+    //
+    /// ct is not required to have clean carries
+    pub fn unchecked_is_odd<T>(&self, ct: &T) -> BooleanBlock
+    where
+        T: IntegerRadixCiphertext,
+    {
+        ct.blocks().first().map_or_else(
+            || {
+                // Interpret empty as being 0, which is not odd
+                self.create_trivial_boolean_block(false)
+            },
+            |first_block| {
+                let result = self.key.unchecked_scalar_bitand(first_block, 1);
+                BooleanBlock::new_unchecked(result)
+            },
+        )
+    }
+
+    /// Returns an encryption of true if the value is even
+    pub fn smart_is_even_parallelized<T>(&self, ct: &mut T) -> BooleanBlock
+    where
+        T: IntegerRadixCiphertext,
+    {
+        // Even though we don't need to worry about carries,
+        // we still clean them
```

Comment:
Yes, but i did that because if we don't clean, we would not need the `&mut` in the input so as options we
* can add an #[allow]
* or make the smart take a `&`,
* or have no smart version
* or change the check in the if be block.carries_are_full()